hadoop-common-project/hadoop-common/src/main/java/io/hops/security/ServiceJWTManager.java

```diff
@@ -0,0 +1,592 @@
+  private final CountDownLatch waiter;
+  
+  private Configuration sslConf;
+  private URL host;
+  private URL serviceJWTRenewPath;
+  private URL serviceJWTInvalidatePath;
+  private long serviceJWTValidityPeriodSeconds;
+  private final AtomicReference<String> masterToken;
+  private LocalDateTime masterTokenExpiration;
+  
+  private String[] renewalTokens;
+  
+  public ServiceJWTManager(String name) {
+    super(name);
+    jsonParser = new GsonBuilder()
+        .setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
+        .setDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
+        .create();
+    this.masterToken = new AtomicReference<>();
+    this.tokenRenewer = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+        .setNameFormat("JWT renewer thread")
+        .setDaemon(true)
+        .build());
+    waiter = new CountDownLatch(1);
+  }
+  
+  public void setHTTPClient(CloseableHttpClient client) {
+    this.httpClient = client;
+  }
+  
+  @Override
+  protected void serviceInit(Configuration conf) throws Exception {
+    host = new URL(conf.get(CommonConfigurationKeysPublic.HOPS_HOPSWORKS_HOST_KEY,
+        "https://127.0.0.1"));
+    sslConf = new Configuration(false);
+    sslConf.addResource(conf.get(SSLFactory.SSL_SERVER_CONF_KEY, "ssl-server.xml"));
+    loadMasterJWT();
+    loadRenewalJWTs();
+    serviceJWTValidityPeriodSeconds = conf.getTimeDuration(
+        CommonConfigurationKeysPublic.JWT_MANAGER_MASTER_TOKEN_VALIDITY_PERIOD,
+        CommonConfigurationKeysPublic.DEFAULT_JWT_MANAGER_MASTER_TOKEN_VALIDITY_PERIOD,
+        TimeUnit.SECONDS);
+    if (serviceJWTValidityPeriodSeconds == 0) {
+      serviceJWTValidityPeriodSeconds = 30L;
+    }
+    String serviceJWTRenewPathConf = conf.get(CommonConfigurationKeysPublic.JWT_MANAGER_SERVICE_RENEW_PATH,
+        CommonConfigurationKeysPublic.DEFAULT_JWT_MANAGER_SERVICE_RENEW_PATH);
+    serviceJWTRenewPath = new URL(host, serviceJWTRenewPathConf);
+    String serviceJWTInvalidatePathConf = conf.get(CommonConfigurationKeysPublic.JWT_MANAGER_SERVICE_INVALIDATE_PATH,
+        CommonConfigurationKeysPublic.DEFAULT_JWT_MANAGER_SERVICE_INVALIDATE_PATH);
+    if (!serviceJWTInvalidatePathConf.endsWith("/")) {
+      serviceJWTInvalidatePathConf = serviceJWTRenewPathConf + "/";
+    }
+    serviceJWTInvalidatePath = new URL(host, serviceJWTInvalidatePathConf);
+    super.serviceInit(conf);
+  }
+  
+  private void createLockFileIfMissing() throws IOException {
+    Path lockFile = getLockFilePath();
+    if (!lockFile.toFile().exists()) {
+      FileChannel fc = FileChannel.open(lockFile, StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+      FileLock fl = fc.tryLock();
+      if (fl == null) {
+        return;
+      }
+      try {
+        Files.setPosixFilePermissions(lockFile, LOCK_FILE_PERMISSIONS);
+      } finally {
+        fl.release();
+      }
+    } else {
+      Files.setPosixFilePermissions(lockFile, LOCK_FILE_PERMISSIONS);
```

Comment:
The problem I am trying to solve here is synchronization among different JVMs running on the same machine. The `ServiceJWTManager` is responsible for renewing the JWT in `ssl-server.xml` file which is used to authenticate to Hopsworks.

In a setup where a DN is collocated with a RM, both they will try to renew the JWT as both services need to communicate with Hopsworks. This can end-up with one service invalidating the token that the other service has in-memory. So we have to synchronize them outside a JVM by this _lock_ file.

The main usage of this _lock_ file is in the `TokenRenewer` thread. Periodically it tries to acquire the lock on that file. If it succeeds then it updates the JWTs in `ssl-server.xml`, otherwise it means there is another process responsible for updating and it just refreshes the in-memory state.

Regarding this comment, if the file does **not** exist, take an exclusive lock on the file to avoid another process which is starting at the same time to create the file. If the file is there then just fix the permissions so the renewer thread can read it later. So there is no need for a lock if the file already exists.